File: Assignment/longestCommonPrefix.hs
--Name: Ethan Sharkey
--Student No.: 17355756

-- Used to test Efficiency
--------------------------------------------------------------------------------------------
--module Main where
--main = putStrLn (longestPrefix ["interview", "interrupt", "integrate", "intermediate"])
--------------------------------------------------------------------------------------------
-- Test Case
--["interview", "interrupt", "integrate", "intermediate"]

longestPrefix :: (Eq words) => [[words]] -> [words]
longestPrefix = foldl1 longestCommonPrefix -- implementing foldl1 on longestcommonprefix


longestCommonPrefix :: (Eq letter) => [letter] -> [letter] -> [letter] -- takes in two letters and returns a letter or whitespace

longestCommonPrefix _ [] = [] -- two base cases
longestCommonPrefix [] _ = []

longestCommonPrefix (x : xs) (y : ys) -- calling function on the two lists
    | x == y = x : longestCommonPrefix xs ys -- if the two letters equal each other, add to prefix, otherwise, stop. signifies end of prefix
    | otherwise = []

File: Assignment/longestCommonPrefix.pl
% Name: Ethan Sharkey
%Student No.: 17355756

/**Used as test case = ["interview", "interrupt", "integrate", "intermediate"]*/

:- set_prolog_flag(double_quotes, chars).  % "abc" = [a,b,c]

prefix_of(Prefix, List) :- % Appends the Letters intvolved in the prefix to a list
   append(Prefix, _, List).

toString(Str,Longest):-                  %Takes a list of atoms and converts it to a String
    atom_chars(Str,Longest).             %Using inbuilt atom_chars function

longestCommonprefix(Prefix, List) :-
   findall(Letters, maplist(prefix_of(Letters),List), ListOfAllPrefixes), % Between findall and maplist, it finds all prefixes possible  
   last(ListOfAllPrefixes, LongestPrefix),                                % and adds them to a list. last gets the longest element in the list created above.
   toString(Prefix, LongestPrefix).                                       % This is because the last element is the longest. toString creates it as a string.